knox/crypto.py

```diff
@@ -14,10 +14,15 @@ def create_token_string() -> str:
 
 def make_hex_compatible(token: str) -> bytes:
     """
-    We need to make sure that the token, that is send is hex-compatible.
-    When a token prefix is used, we cannot guarantee that.
+    Ensure a token, which may contain a TOKEN_PREFIX, is hex-compatible before hashing.
     """
-    return binascii.unhexlify(binascii.hexlify(bytes(token, 'utf-8')))
+    try:
+        # use the original method for hashing a token pre v5.0.*
+        # this ensure tokens generated in v4.2 will remain valid in v5.1+
+        return binascii.unhexlify(token)
+    except (binascii.Error, ValueError):
+        # if a token has a prefix, encode it so that it's hex-compatible and can be hashed
+        return binascii.hexlify(token.encode('utf-8'))
```

Comment:
Is there a reason for not keeping the 5.0 implementation?

```suggestion
        return binascii.unhexlify(binascii.hexlify(bytes(token, 'utf-8')))
```

Reply:
No particular reason, actually. Instead, `binascii.unhexlify(binascii.hexlify(bytes(token, 'utf-8')))` could be simplified to just `bytes(token, 'utf-8')` - it'll produce the same thing.

And actually, if we take this approach, we can at least preserve partial backwards compatibility - any tokens generated with a _prefix_ in 5.0.* will stick work when upgrading to 5.1.* (not sure if that's helpful, but it might be slightly better?)

Reply:
I think that is preferable but how annoying that we're still left with one breaking case still. I think we need guidance from the maintainers here about what they would be happy with. I agree that checking tokens twice does not feel great.